RFC: Documentation philosophy for ODP projects (#44)
OpenDevicePartnership/governance: rfc/0000-documentation-philosophy.md

- **GitHub Pages** publishes static sites directly from a GitHub repository,
  typically via a GitHub Actions workflow that builds the mdBook on push and
  deploys the rendered output.

## Goals

1. Establish a single, consistent documentation philosophy across all ODP
   projects, expressed as two clear principles:
   - No project-specific documentation lives in the `documentation`
     repository.
   - Every ODP project has its own project-level mdBook.
2. Give each project a first-class, independently owned documentation
   surface that can evolve with the project.
3. Standardize the documentation toolchain on **mdBook + GitHub Pages** so
   contributors, reviewers, and readers have a consistent experience across
   projects.
4. Preserve a clear home for **organization-level** documentation (charter,
   governance, cross-project standards, the RFC index) that is not
   project-specific.
5. Migrate existing project-specific documentation currently in the
   `documentation` repository (Secure EC, Standardized EC Services) into
   the corresponding project's mdBook without loss of content or history
   where practical.

## Requirements

> 1. **No project-specific docs in the `documentation` repository.** The

Reviewer: My concern here is there is no "project level" repo. Some projects span multiple repos. For those are we proposing we have one high level repo that represents each project? Or do we keep the documentation for those projects at the documentation repo level?

Author: @philgweber - yes, its up to the project lead to decide which of the repo's to use as the place to host the project mdBook out of.

The migration is performed **per project** and is expected to run
incrementally. For each project currently documented in the `documentation`
repository (initially: Secure EC and Standardized EC Services), the owning
maintainers:

1. **Stand up mdBook in a project repository.**
   - Add a `docs/` directory containing `book.toml`, `src/SUMMARY.md`, and
     an initial `src/` tree.
   - Add a GitHub Actions workflow that builds the mdBook and deploys it to
     GitHub Pages on push to the default branch.
   - Enable GitHub Pages on the project repository.
2. **Move content.** Copy the project's documentation from the
   `documentation` repository into `docs/src/` in the project repository,
   preserving structure where it still makes sense. Update internal links
   to be relative to the new location.
3. **Publish.** Land the new mdBook, verify the published GitHub Pages
   site, and confirm the `SUMMARY.md` covers the migrated content.

Once all previously centralized project docs have been migrated, the
`documentation` repository's top-level index/landing page is updated to
enumerate ODP projects and link to each project's published mdBook.

## Unresolved Questions

> - **Migration timeline.** This RFC does not fix a completion date. Should

Reviewer 2: We have been talking about this for a while. To prompt some action, we should create backlog of issues to do the changes and then assign it to an iteration.

1. Top-level README for the ODP GitHub highlighting the top-level repo for each project
2. Move project specific documentation out of the documentation repo, change documentation repo to be more top-level overview.
3. For project lead and owner to create their documentation and publish as gh-pages for centralized documentation to point to

Author: @jerrysxie - I have these items in our backlog loop. We can prioritize appropriately when we do sprint planning.

# RFC: Documentation Philosophy for ODP Projects

This RFC proposes that Open Device Partnership (ODP) adopt a documentation
philosophy built on two principles:

1. **No project-specific documentation lives in the `documentation`
   repository.**
2. **Each ODP project has its own project-level [mdBook][mdbook]**, owned
   and maintained by that project and published via GitHub Pages.

The `documentation` repository continues to exist, but its role narrows to
hosting **organization-level** content (charter, governance, cross-project
standards, the RFC index) and to acting as a landing page that links out to
each project's mdBook.

## Change Log

- 2026-07-21: Initial RFC draft created.
- 2026-07-27: Reframed the RFC around two explicit principles ("no
  project-specific docs in the documentation repo" and "each ODP project
  has a project-level mdBook") rather than a comparison between two
  existing approaches.

## Motivation

ODP projects today handle documentation inconsistently. Some projects, such
as Secure EC and Standardized EC Services, keep their documentation in the

> shared `documentation` repository. Others, such as Patina, keep their

Reviewer 3: For clarity, that "`documentation`" repo here means the ODP org-level repo.

```suggestion
ODP projects today handle documentation inconsistently. Some projects, such
as Secure EC and Standardized EC Services, keep their documentation in the
ODP org-level `documentation` repository. Others, such as Patina, keep their
```

documentation with the project as a project-level mdBook. The two
principles proposed by this RFC resolve that inconsistency.

> The current split causes real friction:

Reviewer 3: This section still focuses heavily on the co-location aspect. It that's being treated as a suggestion, can you update this to emphasize the two main points ("(1) project mdbook not in ODP-org documentation repo and (2) each ODP project has a project-wide mdbook")?

Then, you can move the doc & code co-location section to a suggestion/guidance area.

Author: Tried to clarify this even more and have listed it as a requirement that we are not dictating specific location of project level mdBooks.

   project-specific.
5. Migrate existing project-specific documentation currently in the
   `documentation` repository (Secure EC, Standardized EC Services) into
   the corresponding project's mdBook without loss of content or history
   where practical.
6. Keep the migration cost low and incremental, so projects are not blocked
   on a big-bang cutover.

## Requirements

1. **No project-specific docs in the `documentation` repository.** The
   `documentation` repository hosts only organization-level content:
   charter, governance, cross-project standards and conventions, the RFC
   index, and a landing page that links out to each project's published
   mdBook. Project-specific technical documentation does not belong there.
2. **Every ODP project has a project-level mdBook.** Each project owns and
   maintains its own mdBook (Markdown + `SUMMARY.md`, built with `mdbook`)
   covering that project's documentation. NOTE: this does not mean that
   every repository in the organization must have an `mdbook`.
3. **Project mdBooks live in the project's repository** under a
   conventional in-repo path (default: `docs/`), so documentation is

>    owned, reviewed, and versioned alongside the project.

Reviewer 3: ```suggestion
3. **Project mdBooks live in an appropriate project repository** under a
   conventional in-repo path (default: `docs/`), so documentation is
   owned, reviewed, and versioned alongside the project.
```

4. **GitHub Pages publishing.** Each project publishes its rendered mdBook
   via GitHub Pages from the project repository, using a GitHub Actions

>    workflow that builds on push to the project's default branch.

Reviewer 3: ```suggestion
4. **GitHub Pages publishing.** Each project publishes its rendered mdBook
   via GitHub Pages from a project repository, using a GitHub Actions
   workflow that builds on push to the project's default branch.
```

   in the `documentation` repository with short stubs that link to the new
   canonical location on the project's GitHub Pages site, so existing
   inbound links continue to work.
5. **Announce.** Note the cutover in the project's usual communication
   channels so downstream readers and contributors update their bookmarks.

Once all previously centralized project docs have been migrated, the
`documentation` repository's top-level index/landing page is updated to
enumerate ODP projects and link to each project's published mdBook.

## Unresolved Questions

- **Migration timeline.** This RFC does not fix a completion date. Should
  a target window (e.g. "by the next release cycle") be set for the
  initial two projects, or is best-effort acceptable?

## Prior Art

- **Patina** already co-locates its documentation with its source code and
  publishes it as an mdBook. This RFC generalizes Patina's approach to the
  rest of ODP.
- **Docs-as-code more broadly.** Treating documentation as source that
  lives with the code, is reviewed in the same PRs, and is built by CI is
  a well-established pattern across open-source projects.

> ## Alternatives

Reviewer 3: As noted in another comment, I think that at an organization level, it's reasonable to say that documentation uses mdbook for consistency in documentation management and user navigation across docs.

That does not mean:

- Projects have to have an mdbook in every repo
- Projects have to exclusively use mdbook (docs.rs should be used for code and more technical docs closer to the code)

It does not necessarily require that an individual project co-locate their docs and code. There are noted advantages to doing so, but the org-level requirement to use mdbook as a documentation standard at each workstream level is achievable regardless of that detail. I suggest moving co-location of code & docs to be a suggestion rather than requirement, at least for now.

Reviewer 3: That said, there documentation should be in project-specific repos not in `documentation` which is reserved for org-level documenation.

Author: agreed @makubacki . I will clarify the RFC.

   the RFC index, and a landing page that links out to each project's
   published mdBook. It does not host project-specific technical
   documentation.
6. **Existing centralized docs are migrated.** Documentation currently held
   in the `documentation` repository for Secure EC and Standardized EC
   Services is moved into those projects' repositories as mdBooks. The
   original locations in the `documentation` repository are then replaced
   with short link stubs pointing at the new canonical location.
7. **New projects follow this model on day one.** Any new project onboarded
   under ODP governance stands up its documentation in-repo as an mdBook
   with a GitHub Pages workflow as part of initial setup.
8. **Stable external URLs.** During migration, projects should preserve
   inbound links where reasonable by leaving redirect/link stubs at the
   previous locations in the `documentation` repository.

## Migration Plan

The migration is performed **per project** and is expected to run
incrementally. For each project currently documented in the `documentation`
repository (initially: Secure EC and Standardized EC Services), the owning
maintainers:

> 1. **Stand up mdBook in the project repository.**

Reviewer 2: We should scope this to the top-level repos we want to drive people to. We don't want to do this for every repo.

Reviewer 3: At an organization level, workstreams need to be clearly identified and documented regardless of repo structure. For example, if @dhillonk were to update an ODP level document to point to documentation for each ODP workstream, where would these EC entries link to?

- Patina: https://opendevicepartnership.github.io/patina/
- Secure EC: ?
- Standardized EC Services: ?

Imo, this is more about that high-level view of these workstreams. Not that each repo would require it.

The value is that if someone is interested in ODP and they see these workstreams, they now have a single place to go (from the org-level doc) that describes the entire workstream without having to manually read each repo's readme (which sounds like a lot) and figure out how the repos are connected.

Otherwise, `cargo doc` can document more technical, smaller, and miscellaneous information fine and places like the mdbook can link to its doc.rs pages.

---

In favor of this RFC, in Patina, we found splitting repos to a large degree made routine maintenance have more overhead, GitHub administration more involved/fragile, fractured visitor attention to the overall project, made code refactors more complicated, and keeping documentation accurate more difficult, so we consolidated.

---

How many top-level repos are there? Do any make sense to host the mdbook nearest relevant code for the workstreams mentioned in the RFC (Secure EC & Standardized EC Services)?

Author: +1 to @makubacki 's comment. The intent is not to require each repo to have an mdBook. Only that each top level project should have 1 top level mdBook. From there its the engineering teams decision if they want to document everything about that project in that 1 mdBook or to fork off into different mdBooks / repo's.
